test/distributed/test_device_mesh.py

```diff
@@ -561,6 +561,34 @@ def test_get_item_3d_noncontinuous_slicing(self):
         # Check on the current cp_local_rank, whether the dp mesh tensor is the same.
         self.assertEqual(cp_dp_mesh.mesh[cp_local_rank], dp_mesh.mesh)
 
+    @with_comms
+    def test_flatten_mesh(self):
+        mesh_shape = (2, 2, 2)
+        mesh_dim_names = ("dp", "cp", "tp")
+        mesh_3d = init_device_mesh(
+            self.device_type, mesh_shape, mesh_dim_names=mesh_dim_names
+        )
+
+        # Test flatten contiguous dims
+        dp_cp_mesh = mesh_3d["dp", "cp"]
+        flattened_dp_cp_mesh = dp_cp_mesh._flatten()
+        self.assertEqual(dp_cp_mesh.mesh.flatten(), flattened_dp_cp_mesh.mesh)
+
+        # Test flatten non-contiguous dims
+        dp_tp_mesh = mesh_3d["dp", "tp"]
+        flattned_dp_tp_mesh = dp_tp_mesh._flatten()
+        self.assertEqual(dp_tp_mesh.mesh.flatten(), flattned_dp_tp_mesh.mesh)
+
+        # Test flatten non-contiguous dims with swapping order
+        # We sort the order of the mesh_dim_names in the mesh to flatten based on
+        # order of the mesh_dim_names in the root mesh, since we do not want to create
+        # DeviceMesh([0, 4, 1, 5], mesh_dim_names=('tp_dp',)) or DeviceMesh([2, 6, 3, 7], mesh_dim_names=('tp_dp',)).
+        # No matter the order of slice given, the result should be
```

Comment:
Is there a reason for this ordered behavior? I feel if user want to swap the mesh dim orders, and flatten the swapped order device mesh, we should probably respect that?

Reply:
Say if we have a 3D mesh and we create mesh["dp", "cp"] and mesh["cp", "dp"], the result does look different and the order matters, because of the 2D mesh_dim order. However, if we flatten out these two, the result would both be a 1D mesh with the same global ranks, except that the mesh tensor's order is different and the mesh_dim_names are going to be different.
```
DeviceMesh([0, 4, 1, 5], mesh_dim_names=('dp_cp',))
DeviceMesh([0, 1, 4, 5], mesh_dim_names=('cp_dp',))
```
If we distinguish these two, then we will create two pgs with different local rank result. I am not sure if we want to distinguish these two.

Reply:
I think this would bring some user confusions about the mesh dim order, If we don't want to distinguish these two, we should simply not allowing reordered mesh dim slicing (i.e. we only allow submesh slicing happen in a contiguous manner like before, but it could still skip the dims in the middle)

Reply:
Sounds good. I removed the sorting part and the test for flattening with swapping dims in this PR. 
I will put up a separate PR to add the restriction for slicing.